### PR TITLE
Add verification target for BMC engine: check if type conversion truncates a value

### PR DESCRIPTION
**Pull request by pgebal**
Implements a piece of https://github.com/ethereum/solidity/issues/5808.
Verification target for CHC engine will be added in a seperate PR.

### CODE REVIEW

File: scripts/error_codes.py [hunk 1]
```diff
@@ -207,6 +207,7 @@ def examine_id_coverage(top_dir, source_id_to_file_names, new_ids_only=False):
         "2961", # SMTChecker, covered by CL tests
         "6240", # SMTChecker, covered by CL tests
         "9576", # SMTChecker, covered by CL tests
+        "9371", # SMTChecker, typeTruncationConversion target
```
Comment (pgebal) [inline scripts/error_codes.py:210]: This is *might happen* version, I'm not happy about this. I do not know how to come up with a test that prints the *might* version.

File: test/libsolidity/smtCheckerTests/types/array_literal_1.sol [hunk 2]
```diff
@@ -11,3 +11,4 @@ contract C
 // SMTEngine: all
 // ----
 // Info 1391: CHC: 6 verification condition(s) proved safe! Enable the model checker option "show proved safe" to see all of them.
+// Info 6002: BMC: 1 verification condition(s) proved safe! Enable the model checker option "show proved safe" to see all of them.
```
Comment (pgebal) [inline test/libsolidity/smtCheckerTests/types/array_literal_1.sol:14]: no new line at the end of file

File: libsolidity/formal/BMC.cpp [hunk 3]
```diff
@@ -701,6 +708,55 @@ void BMC::visitAddMulMod(FunctionCall const& _funCall)
...
+	auto const expressionType = _funCall.expression().annotation().type;
+	auto const resultType = dynamic_cast<TypeType const&>(*expressionType).actualType();
+
+	auto symbArg = expr(*argument, funCallType);
+
+	// do not warn about address(contract)
+	if (smt::isAddress(*resultType) && dynamic_cast<ContractType const*>(argType))
+		return;
+
+	if (smt::isStringLiteral(*argType) && smt::isFixedBytes(*resultType))
```
Comment (pgebal) [inline libsolidity/formal/BMC.cpp:725]: no consistency in conditions; first check resultType then argType

Comment (blishko) [reply libsolidity/formal/BMC.cpp:725]: What do you mean?

File: libsolidity/formal/BMC.cpp [hunk 4]
```diff
@@ -701,6 +708,55 @@ void BMC::visitAddMulMod(FunctionCall const& _funCall)
...
+		return;
+
+	ArrayType const* arrayType = dynamic_cast<ArrayType const*>(argType);
+
+	// no support for array slice
+	if (dynamic_cast<ArraySliceType const*>(argType))
+		return;
+
+	if (arrayType && arrayType->isByteArrayOrString() && smt::isFixedBytes(*resultType))
```
Comment (blishko) [inline libsolidity/formal/BMC.cpp]: We do not need `arrayType` outside of the corresponding `if` block, no?

```suggestion
	// no support for array slice
	if (dynamic_cast<ArraySliceType const*>(argType))
		return;

	if (ArrayType const* arrayType = dynamic_cast<ArrayType const*>(argType); arrayType && arrayType->isByteArrayOrString() && smt::isFixedBytes(*resultType))
```

**Review by blishko — COMMENTED** [review]
Looks good, I will have to dig a bit deeper into the issue.
There is no way for a user to express in Solidity that they are aware about possible truncation and are fine with it? So we would not warn in such cases?

Looking at the tests, I am thinking that maybe this target should not be checked by default, but only when explicitly required? It seems quite verbose, meaning that conversions seem to happen often.

File: libsolidity/formal/BMC.cpp (hunk 3 above)
Comment (pgebal) [reply libsolidity/formal/BMC.cpp:725]: just a note to myself, to be consistent  with the previous `if` statement.

**Comment by pgebal** [comment]
> Looks good, I will have to dig a bit deeper into the issue. There is no way for a user to express in Solidity that they are aware about possible truncation and are fine with it? So we would not warn in such cases?

I don't know about that. I'll ask.


> Looking at the tests, I am thinking that maybe this target should not be checked by default, but only when explicitly required? It seems quite verbose, meaning that conversions seem to happen often.

Good idea. I'll fix it.

**Comment by pgebal** [comment]
Narrowing casts are now explicit in solidity. Due to this reason we decided that verification targets that check truncation are no longer needed. It's on a developer to watch out for potential issues if explicit casts are used.

**Pull request closed without merging**